libsolidity/analysis/PostTypeContractLevelChecker.cpp

```diff
@@ -101,29 +102,64 @@ void PostTypeContractLevelChecker::checkStorageLayoutSpecifier(ContractDefinitio
 	}
 
 	auto const* baseSlotExpressionType = type(baseSlotExpression);
+	auto const* integerType = dynamic_cast<IntegerType const*>(baseSlotExpressionType);
 	auto const* rationalType = dynamic_cast<RationalNumberType const*>(baseSlotExpressionType);
-	if (!rationalType)
+	if (
+		!integerType &&
+		!rationalType
+	)
 	{
+		std::string errorMsg = "The base slot of the storage layout must evaluate to an integer.";
+		if (dynamic_cast<AddressType const*>(baseSlotExpressionType))
+			errorMsg += " The type of the supplied expression is address.";
+		else if (auto const* fixedBytesType = dynamic_cast<FixedBytesType const*>(baseSlotExpressionType))
+			errorMsg += fmt::format(
+				" The type of the supplied expression is bytes{}.",
+				fixedBytesType->numBytes()
+				)
+			;
```

Comment:
I'd quote the type name because otherwise it looks like some grammatical error.

The closing parenthesis is also weirdly indented.

```suggestion
			errorMsg += " The type of the supplied expression is 'address'.";
		else if (auto const* fixedBytesType = dynamic_cast<FixedBytesType const*>(baseSlotExpressionType))
			errorMsg += fmt::format(
				" The type of the supplied expression is 'bytes{}'.",
				fixedBytesType->numBytes()
			);
```